### PR TITLE
Refactoring ABTests

### PR DESCRIPTION
**Pull request by TBonnin**
The goal is to make the creation of AB tests (aka Experiment) easier and less subject to mistakes.
Creating a Experiment would only require adding the following code in Experiments.scala.
An experiment is assigned to a existing participation group picked based on the size of the audience you want to assign to the experiment
```
object MyExperiment extends Experiment(
  name = "my-experiment",
  description = "A slice of the audience will participate in my
experiment"",
  owners = Seq(Owner.withGithub("tbonnin")),
  sellByDate = new LocalDate(2062, 2, 1),
  participationGroup = Perc1A
)
```
That's it! 😱 

Requests would be bucketed in [fixed groups in the VCL](https://github.com/guardian/fastly-edge-cache/pull/804) and frontend apps
would use the Vary response header to signify multiple variants exist for the same request
Groups other than the 50% do NOT overlap to ensure experiments don't influence each other.

This PR also automatically creates an opt-in/out endpoint for each experiment

Another advantage of this solution is cache would only be splitted if the response does depends on a experiment. (ie: no experiment used to generate a page = no split)

Disadvantages of this approach:
- No support for multi-variant tests (more than 2 variants). Very few experiments requires it, for instance none of current experiments have multi-variants
- when running a 1 way experiment (ramp-up or opt-in only) the cache is still splitted in 3 (`variant`, `control` and `none`) rather than 2 and 1 bucket is "wasted"
- If you are opt in a experiment and the participation group for this experiment is being re-assigned to another experiment you will automatically be opt in this new experiment. I have added a `opt/reset` endpoint that would help getting back to a clean state.

TODO: check if/how server side experiments are being used client side

Feeback welcome. Let me know what you think

## What is the value of this and can you measure success?
Modifying the VCL when creating/updating an AB test is currently very error-prone. I personally made 2 mistakes recently when doing so. One mistake resulted in a facia outage because cache key was modified for most of the content. The other resulted in splitting the cache unnecessarily 

## Screenshots
https://www.theguardian.com/discover-cool-canada/2017/sep/25/can-you-tell-a-canadien-from-a-canuck-test-your-canadian-sport-knowledge-quiz
screen shot 2017-11-23 at 17 46 17

## Tested in CODE?
Not yet

### CODE REVIEW

File: common/app/experiments/ExperimentsDefinition.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  def getJavascriptConfig(implicit request: RequestHeader): String =
+    allExperiments
+      .filter(_.isParticipating)
+      .toSeq.sortBy(_.name)
+      .map(experiment => s""""${CamelCase.fromHyphenated(experiment.name)}"""")
+      .mkString(",")
+
+  def isParticipating(experiment: Experiment)(implicit request: RequestHeader): Boolean = {
+    LookedAtExperiments.addExperiment(experiment)
```
Comment (nicl) [inline common/app/experiments/ExperimentsDefinition.scala]: Worth calling out the side effect in a comment here?

File: common/test/common/ExperimentsTest.scala [hunk 2]
```diff
@@ -0,0 +1,132 @@
+package experiments
+
+import conf.switches.Owner
+import org.joda.time.LocalDate
+import org.scalatest.{FlatSpec, Matchers}
+import test.TestRequest
+import ParticipationGroups._
+import play.api.mvc.RequestHeader
+
+class ExperimentsTest extends FlatSpec with Matchers {
```
Comment (rich-nguyen) [inline common/test/common/ExperimentsTest.scala]: Can you add a test to check for duplicate participation groups? https://github.com/guardian/frontend/pull/8886/files#diff-09e577872825ce2f34f6dab651087231R13

File: common/app/experiments/Experiments.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  )
+  implicit val canCheckAB = new CanCheckExperiment(this)
+}
+
+object CommercialClientLogging extends Experiment(
+  name = "commercial-client-logging",
+  description = "A slice of the audience who will post their commercial js performance data",
+  owners = Seq(Owner.withGithub("rich-nguyen")),
+  sellByDate = new LocalDate(2018, 2, 1),
+  participationGroup = Perc1A
```
Comment (rich-nguyen) [inline common/app/experiments/Experiments.scala]: Ah, the [good old days](https://github.com/guardian/frontend/pull/8886). I think this is approach is much less error prone. I wonder if you should allow participation groups to be a `List` of groups, to allow the most flexibility to build the right audience size?

File: common/test/common/ExperimentsTest.scala (hunk 2 above)
Comment (TBonnin) [reply common/test/common/ExperimentsTest.scala]: I thought about it but I wasn't not sure if we should absolutely prevent duplicate participation groups. There is nothing wrong using the same group in theory if you know what you are doing.
However I agree with you we can prevent it for now. It might actually never be an issue

File: common/app/experiments/Experiments.scala (hunk 3 above)
Comment (TBonnin) [reply common/app/experiments/Experiments.scala]: The thing is this implementation is based on the assumption a request is bucketed into only 2 groups in Fastly VCL (`variant` and `control`). https://github.com/guardian/fastly-edge-cache/pull/804/files
Making the participation group dynamic would require the VCL to be aware of it which is impossible without modifying the VCL when creating the experiment (which is the thing this PR is absoluterly trying to prevent)

Comment (rich-nguyen) [reply common/app/experiments/Experiments.scala]: So how does multi-variant work, say an A-B-C test? book out two groups?

Comment (TBonnin) [reply common/app/experiments/Experiments.scala]: It doesn't support multi-variant. None of current experiments requires it and I can't remember last time we needed it.
You can simulate a 3 way multi-variant test with 2 groups assigned to participation groups of the same size. The drawback woule be that cache is partitioned 4 times rather than 3 and one bucket is "wasted". However it happens so rarely that I think we can live with it.

File: common/app/experiments/LookedAtExperiments.scala [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package experiments
+
+import play.api.libs.typedmap.TypedKey
+import play.api.mvc.RequestHeader
+
+import scala.collection.mutable
+
+object LookedAtExperiments {
+
+  private val attrKey: TypedKey[mutable.Set[Experiment]] = TypedKey[mutable.Set[Experiment]]("lookedAtExperiments")
```
Comment (QuarpT) [inline common/app/experiments/LookedAtExperiments.scala]: Might be dangerous using a mutable.Set for multi threaded behaviour. Could use a Java ConcurrentHashMap and convert that to a Scala Set when reading.

File: common/app/mvt/MultiVariateTesting.scala [hunk 5]
```diff
@@ -1,193 +0,0 @@
-package mvt
```
Comment (QuarpT) [inline common/app/mvt/MultiVariateTesting.scala]: 👍

File: common/app/http/Filters.scala [hunk 6]
```diff
@@ -71,6 +72,23 @@ class SurrogateKeyFilter(implicit val mat: Materializer, executionContext: Execu
   }
 }
 
+class ExperimentsFilter(implicit val mat: Materializer, executionContext: ExecutionContext) extends Filter {
+
+  override def apply(nextFilter: (RequestHeader) => Future[Result])(request: RequestHeader): Future[Result] = {
```
Comment (QuarpT) [inline common/app/http/Filters.scala]: could split this function for readability

Comment (TBonnin) [reply common/app/http/Filters.scala]: done https://github.com/guardian/frontend/pull/18320/files#diff-63195b61cf92b9a86ecd3f030f7da391R84

**Comment by gustavpursche** [comment]
In case you are going to merge this soon, please note that another `mvt.` check was introduced on master: https://github.com/guardian/frontend/commit/88edcea5b2bc40072d0170fbc5807a45930acfd5

**Comment by TBonnin** [comment]
Already in there https://github.com/guardian/frontend/pull/18320/files#diff-89c79c0e096c91961ee0ad11bd007482R65 :D

**Comment by gustavpursche** [comment]
@TBonnin But not the actually check `@if(mvt.Garnett.isParticipating)`, no?

**Comment by TBonnin** [comment]
Indeed. The compiler will complain though. Thanks for letting me know.
